thread_pthread.c

```diff
@@ -101,6 +101,78 @@
 #  endif
 #endif
 
+static gvl_hook_t * rb_gvl_hooks = NULL;
+static pthread_rwlock_t rb_gvl_hooks_rw_lock = PTHREAD_RWLOCK_INITIALIZER;
+
+gvl_hook_t *
+rb_gvl_event_new(void *callback, rb_event_flag_t event) {
+    gvl_hook_t *hook = ALLOC_N(gvl_hook_t, 1);
+    hook->callback = callback;
+    hook->event = event;
+
+    if (pthread_rwlock_wrlock(&rb_gvl_hooks_rw_lock)) {
+        // TODO: better way to deal with error?
```

Comment:
It looks like the other pthread functions in Ruby hard crash with `rb_bug`. The documentation says that it can only error `pthread_rwlock_wrlock` can return is

> EDEADLK The current thread already owns the read-write lock for writing or reading.

This shouldn't ever happen I think, so maybe `rb_bug` should be used here too.

Reply:
Oh! That's great info! Thanks!